clang/test/Analysis/nullability.c

```diff
@@ -10,3 +12,42 @@ void block_arity_mismatch() {
   void(^b)() = ^(int a, int b) { };
   b(1);  // no-crash expected-warning {{Block taking 2 arguments is called with fewer (1)}}
 }
+
+int *nonnull_return_annotation_indirect() __attribute__((returns_nonnull));
+int *nonnull_return_annotation_indirect() {
+  int *x = 0;
+  return x; // expected-warning {{Null returned from a function that is expected to return a non-null value}}
+}
+
+int *nonnull_return_annotation_direct() __attribute__((returns_nonnull));
+int *nonnull_return_annotation_direct() {
+  return 0; // expected-warning {{Null returned from a function that is expected to return a non-null value}}
+} // expected-warning@-1 {{null returned from function that requires a non-null return value}}
+
+int *nonnull_return_annotation_assumed() __attribute__((returns_nonnull));
+int *nonnull_return_annotation_assumed(int* ptr) {
+  if (ptr) {
+    return ptr;
+  }
+  return ptr; // expected-warning {{Null returned from a function that is expected to return a non-null value}}
+}
+
+int *produce_nonnull_ptr() __attribute__((returns_nonnull));
+
+__attribute__((returns_nonnull))
+int *cannot_return_null() {
+  int *x = produce_nonnull_ptr();
+  if (!x) {
+    clang_analyzer_warnIfReached();
+    // Incorrect: expected-warning@-1 {{REACHABLE}}
+    // According to produce_nonnull_ptr contract, x cannot be null.
```

Comment:
```suggestion
    // expected-warning@-1 {{REACHABLE}}
    // TODO: This warning is a false positive, according to the contract of
    // produce_nonnull_ptr, x cannot be null.
```
(Just bikeshedding.)

By the way, do you have plans to eliminate this false positive (perhaps in a follow-up commit)?

Reply:
We've entered it as a Jira ticket: https://sonarsource.atlassian.net/browse/CPP-5614
But we have no immediate plans to implement it.